### PR TITLE
Fix option --libs output library path

### PR DESCRIPTION
**Pull request by JackBoosY**
The `dep` does not consider the cmake generator expression and the absolute path.
We may meet some different situation:
1. `dep` was already handled like `-pthread`:
`dep_name` should be `${dep}`.
2. `dep` is system library like `dl` / `m`:
`dep_name` should be `-l${dep}`
3. `dep` is `$<$<CONFIG:DEBUG>:LIB_PATH>` or `$<$<NOT:$<CONFIG:DEBUG>>:LIB_PATH>` that from `LIB-config.cmake`:
`dep_name` should be `${LIB_PATH}`.
4. `dep` is an absolute path like `/usr/home/lib.a` or `/usr/home/lib.so` etc:
`dep_name` should be `${dep}`.

Related: https://github.com/microsoft/vcpkg/pull/18580

### CODE REVIEW

**Comment by vadz** [comment]
@MaartenBent could you please review this? TIA!

File: build/cmake/config.cmake [hunk 1]
```diff
@@ -40,16 +40,26 @@ macro(wx_get_dependencies var lib)
                     get_target_property(dep_name ${dep} OUTPUT_NAME)
                 endif()
             else()
-                get_filename_component(dep_name ${dep} NAME)
+                # For the vaule like $<$<CONFIG:DEBUG>:LIB_PATH>
```
Comment (MaartenBent) [inline build/cmake/config.cmake:43]:
```suggestion
                # For the value like $<$<CONFIG:DEBUG>:LIB_PATH>
```

File: build/cmake/config.cmake [hunk 2]
```diff
@@ -40,16 +40,26 @@ macro(wx_get_dependencies var lib)
...
+                # Or $<$<NOT:$<CONFIG:DEBUG>>:LIB_PATH>
+                string(REGEX REPLACE "^.+>:(.+)>$" "\\1" dep_name ${dep})
+                if (NOT dep_name)
+                    set(dep_name ${dep})
+                endif()
```
Comment (MaartenBent) [inline build/cmake/config.cmake:48]: Why did you remove `get_filename_component(dep_name ${dep} NAME)`? We need this to get the library name.
With this change it will be `/Applications/Xcode.app/Contents/..../MacOSX11.3.sdk/usr/lib/libz.tbd` instead of  `-lz`, or `/usr/lib/x86_64-linux-gnu/libX11.so` instead of `-lX11`.

**Review by MaartenBent — CHANGES_REQUESTED** [review]
This breaks determining the external libraries, see my inline comments.

File: build/cmake/config.cmake (hunk 2 above)
Comment (JackBoosY) [reply build/cmake/config.cmake:48]: For some port use `*Config.cmake` or `*-config.cmake`, the `INTERFACE_LINK_LIBRARIES` may contains cmake generator expression like `$<$<NOT:$<CONFIG:DEBUG>>:LIB_PATH>`.
If we use `get_filename_component(dep_name ${dep} NAME)`, the value of `dep_name` is `NAME>`.

Comment (MaartenBent) [reply build/cmake/config.cmake:48]: But can you use the regex to extract the library from the generator expression, and then use `get_filename_component` on this?

Comment (JackBoosY) [reply build/cmake/config.cmake:48]: I don't know why you use the file name directly here.
If the dependency library is not in system like _/usr/lib_, the compiler will fail to search this library.

Comment (MaartenBent) [reply build/cmake/config.cmake:48]: I try to generate the same output as `configure` does when generating `wx-config`. and it only has the library names.
If it is in a different path, maybe we have to specify `-Lpath/to/lib` before the library. Or require the user to have it in `PATH`.

Comment (JackBoosY) [reply build/cmake/config.cmake:48]: Yes, we must choose one way to solve this issue.

**Comment by talregev** [comment]
@JackBoosY @MaartenBent Any news?

**Comment by MaartenBent** [comment]
I can't really work on this, since on the build systems I can test this on (Windows MSVC & MinGW, Ubuntu, macOS) I do not get paths with generator expressions.
Maybe you can give a few examples of such paths?

What about the following solution?
If `dep` contains a generator expression, we extract both the path and library name. In pseudo code:
```
if (dep contains generator expression)
    dep_path = regex to extract path
    dep_name = regex to extract name
else()
    get_filename_component(dep_name ${dep} NAME)
endif()
```
And before adding the library name, we add the library path:
```
if (dep_path)
    wx_string_append(${var} "-L${dep_path} ")
endif()
if(dep_name STREQUAL "libc.so")
...
```

This should hopefully fix your problem. Some other additions, but probably not needed immediately, is keeping track of the added  `dep_path` so they are added only once. And maybe extracting the path in the no-generator-expression case too.

**Comment by Be-ing** [comment]
Thanks for working on this. I hope this gets resolved soon so we can [use vcpkg for Tenacity](https://github.com/tenacityteam/tenacity/pull/228#issuecomment-880113285).

**Comment by talregev** [comment]
Also for our side can get resolve: 
https://github.com/BOINC/boinc/pull/4385

**Review by MaartenBent — APPROVED** [review]
After rethinking this, I suppose it could be merged as it is now.
It will still have some libraries with full path and extension even when it is not required (like I mentioned earlier https://github.com/wxWidgets/wxWidgets/pull/2402#discussion_r658199266), but I suppose it is more important to keep the path in case it is not in a default location.

And if needed, we could add something in the future that strips default locations like `/usr/lib` or the macOS sdk location.

**Comment by vadz** [comment]
Thanks for the changes and for the review, merging it now with just a minor fix for a typo in a comment.
